<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Chaitali Chakraborty">

<title>Loop_analysis.R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="loop-additional-analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="loop-additional-analysis_files/libs/quarto-html/quarto.js"></script>
<script src="loop-additional-analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="loop-additional-analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="loop-additional-analysis_files/libs/quarto-html/anchor.min.js"></script>
<link href="loop-additional-analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="loop-additional-analysis_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="loop-additional-analysis_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="loop-additional-analysis_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="loop-additional-analysis_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Loop_analysis.R</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Chaitali Chakraborty </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="downstream-analysis-of-hichiphic-loops" class="level2">
<h2 class="anchored" data-anchor-id="downstream-analysis-of-hichiphic-loops">Downstream analysis of HiChIP/HiC loops</h2>
<p>This shiort tutorial aims to integrate example chromation loops that you may have acquired through hichipper etc. The files to be used are in the Github page remeseirogrp/3D-genomics-blueprint</p>
</section>
<section id="contents" class="level2">
<h2 class="anchored" data-anchor-id="contents">Contents</h2>
<p>Workflow</p>
<ul>
<li><p>Overlap of loops</p></li>
<li><p>UCSC .interact files</p></li>
<li><p>Annotating loops to genome</p></li>
<li><p>Analysing length of differential loops</p></li>
<li><p>Extracting anchors from loops, for more downstream analysis</p></li>
</ul>
</section>
<section id="load-libraries" class="level2">
<h2 class="anchored" data-anchor-id="load-libraries">Load libraries</h2>
<p>The prior thing to do before this is install alrequired libraries, using requiredPackages.R. Then run this</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({<span class="fu">library</span>(ggplot2)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(stringr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ggpubr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(rstatix)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(circlize)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(devtools)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(RColorBrewer)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(clusterProfiler)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(GenomicRanges)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(rtracklayer)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(BSgenome)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(BSgenome.Hsapiens.UCSC.hg38)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg38.knownGene)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(InteractionSet)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(GenomicInteractions)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(edgeR)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(limma)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(clusterProfiler)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(DESeq2)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(Sushi)</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(diffloop)})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in fun(libname, pkgname): Package 'Sushi' is deprecated and will be removed from Bioconductor
  version 3.4</code></pre>
</div>
</div>
<section id="overlap-of-loops" class="level4">
<h4 class="anchored" data-anchor-id="overlap-of-loops">Overlap of loops</h4>
<p>We want to find out which loops are present specifically in one condition to the other (GB specific loops, hAstro specific loops), and loops that are v present in both cells (common).</p>
<p>This dataframe was obtained from diffloops analysis. You can get a similar data frame from FitHiChIP. Both diffloops and FitHiCHIP can take loops from individual sample loops and merge the anchors, in proximity of 500 bp and recreate loops, that are present in the input samples. The counts indicate, how many times the loop was called for each sample.</p>
<p>FitHiC and HiCompare does similar merging, for HiC loos, It is better that you use an avialbale and trusted tool for merging pf anchors, it makes it easy.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The dataset that we are loading here is a dummy dataset created from the HiChIP data of https://doi.org/10.1038/s41467-023-41919-x</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"~/Desktop/PALS2024_3Dcourse/PALS2024/data_chr15.2_studentsdataframe.txt"</span>) <span class="co">#change your path here</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       chr.x  start.x    end.x chr.y  start.y    end.y hAstro U3028 loopWidth
184974 chr15 17050582 17070200 chr15 17075230 17086156      3     0     20302
184977 chr15 17435126 17441610 chr15 18230127 18231848      0     2    792619
184978 chr15 17457775 17460493 chr15 18230127 18231848      0     2    771853
184985 chr15 18465420 18523463 chr15 19167038 19170276      0     1    674215
184988 chr15 18754402 18770002 chr15 19070204 19108832      0     1    327316
184996 chr15 19794133 19797871 chr15 20183050 20208457      2     0    399751
          mango.FDR      mango.P
184974 9.926786e-04 6.096431e-04
184977 5.180832e-08 1.517130e-08
184978 2.378743e-05 1.058543e-05
184985 3.138288e-05 1.445453e-05
184988 1.265685e-04 6.808408e-05
184996 8.493615e-05 4.390431e-05</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#here you want to create an id for the loops, for tracing purposes</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>loopID <span class="ot">&lt;-</span> <span class="fu">paste</span>(data<span class="sc">$</span>chr.x, data<span class="sc">$</span>start.x, <span class="at">sep =</span> <span class="st">":"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>loopID <span class="ot">&lt;-</span> <span class="fu">paste</span>(data<span class="sc">$</span>loopID, data<span class="sc">$</span>end.x, <span class="at">sep =</span> <span class="st">"-"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>loopID <span class="ot">&lt;-</span> <span class="fu">paste</span>(data<span class="sc">$</span>loopID, data<span class="sc">$</span>chr.y, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>loopID <span class="ot">&lt;-</span> <span class="fu">paste</span>(data<span class="sc">$</span>loopID, data<span class="sc">$</span>start.y, <span class="at">sep =</span> <span class="st">":"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>loopID <span class="ot">&lt;-</span> <span class="fu">paste</span>(data<span class="sc">$</span>loopID, data<span class="sc">$</span>end.y, <span class="at">sep =</span> <span class="st">"-"</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       chr.x  start.x    end.x chr.y  start.y    end.y hAstro U3028 loopWidth
184974 chr15 17050582 17070200 chr15 17075230 17086156      3     0     20302
184977 chr15 17435126 17441610 chr15 18230127 18231848      0     2    792619
184978 chr15 17457775 17460493 chr15 18230127 18231848      0     2    771853
184985 chr15 18465420 18523463 chr15 19167038 19170276      0     1    674215
184988 chr15 18754402 18770002 chr15 19070204 19108832      0     1    327316
184996 chr15 19794133 19797871 chr15 20183050 20208457      2     0    399751
          mango.FDR      mango.P
184974 9.926786e-04 6.096431e-04
184977 5.180832e-08 1.517130e-08
184978 2.378743e-05 1.058543e-05
184985 3.138288e-05 1.445453e-05
184988 1.265685e-04 6.808408e-05
184996 8.493615e-05 4.390431e-05
                                                loopID
184974 chr15:17050582-17070200_chr15:17075230-17086156
184977 chr15:17435126-17441610_chr15:18230127-18231848
184978 chr15:17457775-17460493_chr15:18230127-18231848
184985 chr15:18465420-18523463_chr15:19167038-19170276
184988 chr15:18754402-18770002_chr15:19070204-19108832
184996 chr15:19794133-19797871_chr15:20183050-20208457</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#when we do overlap of loops we need a name or id to trace back the intersection to. This. column gives us this possibility</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>hAstro_loops <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(hAstro <span class="sc">!=</span> <span class="dv">0</span>) <span class="co">#1088</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>hAstro <span class="ot">&lt;-</span> hAstro_loops<span class="sc">$</span>loopID</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>U3028_loops <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(U3028 <span class="sc">!=</span> <span class="dv">0</span>) <span class="co">#3085</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>U3028 <span class="ot">&lt;-</span> U3028_loops<span class="sc">$</span>loopID</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Overlap matrix with loop IDs</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>lt1 <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"hAstro"</span><span class="ot">=</span>hAstro, <span class="st">"U3028"</span><span class="ot">=</span>U3028)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>mat1 <span class="ot">&lt;-</span> <span class="fu">list_to_matrix</span>(lt1)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>ltdf1 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mat1) <span class="co">#a matrix with binaries for TRUE and FALSe on presense and absence of loops</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>mat2 <span class="ot">&lt;-</span> <span class="fu">make_comb_mat</span>(lt1)</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="co">#make an UpSet plot to visualize the overlaps</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="fu">UpSet</span>(mat2, <span class="at">set_order =</span> <span class="fu">c</span>(<span class="st">"hAstro"</span>, <span class="st">"U3028"</span>), <span class="at">top_annotation =</span> <span class="fu">upset_top_annotation</span>(mat2, <span class="at">add_numbers=</span> <span class="cn">TRUE</span>), <span class="at">right_annotation =</span> <span class="fu">upset_right_annotation</span>(mat2, <span class="at">add_numbers=</span><span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#png("data_studentsdataframe_upsetmatrix_UPSETPLOT.png", width = 14, height = 4, #res = 600, units = 'in')</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co">#UpSet(mat2, set_order = c("hAstro", "U3028"), top_annotation = #upset_top_annotation(mat2, add_numbers= TRUE), right_annotation = #upset_right_annotation(mat2, add_numbers=TRUE))</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#par(mar=c(6,5,5,6) + 0.1)</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co">#dev.off()</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>ltdf1<span class="sc">$</span>presence <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(ltdf1[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)])</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co">#let's annotate</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>ltdf1 <span class="ot">&lt;-</span> ltdf1 <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">annotation =</span> <span class="fu">case_when</span>((presence <span class="sc">==</span> <span class="dv">2</span>) <span class="sc">~</span> <span class="st">"common"</span>,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                                                        (presence <span class="sc">==</span> <span class="dv">1</span> <span class="sc">&amp;</span> hAstro <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">~</span><span class="st">"hAstro.specific"</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                                                        (presence <span class="sc">==</span> <span class="dv">1</span> <span class="sc">&amp;</span> U3028 <span class="sc">==</span> <span class="dv">1</span>)<span class="sc">~</span> <span class="st">"GB.specific"</span>))</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ltdf1<span class="sc">$</span>annotation)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
         common     GB.specific hAstro.specific 
            662            2423             426 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#save this file</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#write.csv(ltdf1, "data_studentsdataframe_upsetmatrix_annotated.csv") </span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#now annotate your loop matrix</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>annotation <span class="ot">&lt;-</span> ltdf1<span class="sc">$</span>annotation[<span class="fu">match</span>(data<span class="sc">$</span>loopID, <span class="fu">rownames</span>(ltdf1))]</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">colour =</span> <span class="fu">case_when</span>(data<span class="sc">$</span>annotation <span class="sc">==</span> <span class="st">"common"</span> <span class="sc">~</span> <span class="st">"#a0a0a0"</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>                                                                  data<span class="sc">$</span>annotation <span class="sc">==</span>  <span class="st">"hAstro.specific"</span> <span class="sc">~</span> <span class="st">"#1f78b4"</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>                                                                  data<span class="sc">$</span>annotation <span class="sc">==</span> <span class="st">"GB.specific"</span><span class="sc">~</span><span class="st">"#e31a1c"</span>))</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="co">#these colours are what we aregoing to use to make an interact tract for UCSc</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="co">#write.table(data, "data_studentsdataframe_annotated.txt", sep = "\t", quote = F)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ucsc-.interact-files" class="level4">
<h4 class="anchored" data-anchor-id="ucsc-.interact-files">UCSC .interact files</h4>
<p>Look at the UCSC interact and bigInteract <a href="https://genome.ucsc.edu/goldenPath/help/interact.html" class="uri">https://genome.ucsc.edu/goldenPath/help/interact.html</a>. We are organizing our data frame to UCSC .interact format for visualization of our loops. This is a dummy track for you to look at <a href="https://genome.ucsc.edu/s/chaisan/PALS2024" class="uri">https://genome.ucsc.edu/s/chaisan/PALS2024</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> data[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>)]</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>name <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"Chaitali"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(df), <span class="at">sep =</span> <span class="st">"_"</span>) <span class="co">#you should name your experiment so that you can trace the comaprison, I gave it my name, since it is a dummy file</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>score <span class="ot">&lt;-</span> data<span class="sc">$</span>total_counts</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>value <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fu">log10</span>(data<span class="sc">$</span>mango.FDR)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>value <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df<span class="sc">$</span>value <span class="sc">==</span> <span class="st">"Inf"</span>, <span class="dv">20</span>, df<span class="sc">$</span>value)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(df<span class="sc">$</span>value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#[1] "numeric"</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>exp <span class="ot">&lt;-</span> data<span class="sc">$</span>annotation</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>color <span class="ot">&lt;-</span> data<span class="sc">$</span>colour</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>sourceChrom <span class="ot">&lt;-</span> df<span class="sc">$</span>chr.x</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>sourceStart <span class="ot">&lt;-</span> df<span class="sc">$</span>start.x</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>sourceEnd <span class="ot">&lt;-</span> df<span class="sc">$</span>end.x</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>sourceName <span class="ot">&lt;-</span> <span class="st">"anchor1"</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>sourceStrand <span class="ot">&lt;-</span> <span class="st">"+"</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>targetChrom <span class="ot">&lt;-</span> data<span class="sc">$</span>chr.y</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>targetStart <span class="ot">&lt;-</span> data<span class="sc">$</span>start.y</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>targetEnd <span class="ot">&lt;-</span> data<span class="sc">$</span>end.y</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>targetName <span class="ot">&lt;-</span> <span class="st">"anchor2"</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>targetStrand <span class="ot">&lt;-</span> <span class="st">"+"</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       chr.x  start.x    end.x       name    value             exp   color
184974 chr15 17050582 17070200 Chaitali_1 3.003191 hAstro.specific #1f78b4
184977 chr15 17435126 17441610 Chaitali_2 7.285601     GB.specific #e31a1c
184978 chr15 17457775 17460493 Chaitali_3 4.623652     GB.specific #e31a1c
184985 chr15 18465420 18523463 Chaitali_4 4.503307     GB.specific #e31a1c
184988 chr15 18754402 18770002 Chaitali_5 3.897674     GB.specific #e31a1c
184996 chr15 19794133 19797871 Chaitali_6 4.070907 hAstro.specific #1f78b4
       sourceChrom sourceStart sourceEnd sourceName sourceStrand targetChrom
184974       chr15    17050582  17070200    anchor1            +       chr15
184977       chr15    17435126  17441610    anchor1            +       chr15
184978       chr15    17457775  17460493    anchor1            +       chr15
184985       chr15    18465420  18523463    anchor1            +       chr15
184988       chr15    18754402  18770002    anchor1            +       chr15
184996       chr15    19794133  19797871    anchor1            +       chr15
       targetStart targetEnd targetName targetStrand
184974    17075230  17086156    anchor2            +
184977    18230127  18231848    anchor2            +
184978    18230127  18231848    anchor2            +
184985    19167038  19170276    anchor2            +
184988    19070204  19108832    anchor2            +
184996    20183050  20208457    anchor2            +</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#write.table(df, "data_studentsdataframe_annotated_UCSCinteract.txt", sep = "\t", quote = F)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now take this file to terminal/ shell to trim it so that it can be uploaded to your UCSC browser.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#tail -n +2 data_studentsdataframe_annotated_UCSCinteract.txt &gt; tmpfile</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#less tmpfile</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co">#cut -f 2-19 tmpfile &gt; tmpfile1</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co">#less tmpfile1</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co">#echo -e " track type=interact name="H3K4Me3.HiChIP" description="Chr15 subset" interactDirectional=true maxHeightPixels=200:100:50 visibility=full itemRgb="On" " | cat - tmpfile1  &gt; data_studentsdataframe_annotated_UCSCbrowser.interact</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="annotating-loops-to-genome" class="level4">
<h4 class="anchored" data-anchor-id="annotating-loops-to-genome">Annotating loops to genome</h4>
<p>Genome annotation helps us understand which genomic regions and how many of them are folded in a specific cell or condition with respect to control.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#prepare bedpe files  #8 coloumns - chr.x. start.x, end.x, chr.y, start.y, end.y, name,counts/score/fdr</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co">#astro specific </span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>astro_specific <span class="ot">&lt;-</span> data[data<span class="sc">$</span>annotation <span class="sc">==</span> <span class="st">"hAstro.specific"</span>, ]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(astro_specific)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       chr.x  start.x    end.x chr.y  start.y    end.y hAstro U3028 loopWidth
184974 chr15 17050582 17070200 chr15 17075230 17086156      3     0     20302
184996 chr15 19794133 19797871 chr15 20183050 20208457      2     0    399751
184997 chr15 19853141 19861014 chr15 20406134 20423728      2     0    557853
184998 chr15 19931700 19938878 chr15 20319543 20359651      2     0    404308
184999 chr15 19971427 19974168 chr15 20319543 20359651      2     0    366799
185001 chr15 20027835 20074732 chr15 20183050 20208457      2     0    144470
          mango.FDR      mango.P
184974 9.926786e-04 6.096431e-04
184996 8.493615e-05 4.390431e-05
184997 4.462680e-05 2.146596e-05
184998 7.987363e-05 4.102531e-05
184999 9.928249e-05 5.215215e-05
185001 5.233075e-11 1.134515e-11
                                                loopID      annotation  colour
184974 chr15:17050582-17070200_chr15:17075230-17086156 hAstro.specific #1f78b4
184996 chr15:19794133-19797871_chr15:20183050-20208457 hAstro.specific #1f78b4
184997 chr15:19853141-19861014_chr15:20406134-20423728 hAstro.specific #1f78b4
184998 chr15:19931700-19938878_chr15:20319543-20359651 hAstro.specific #1f78b4
184999 chr15:19971427-19974168_chr15:20319543-20359651 hAstro.specific #1f78b4
185001 chr15:20027835-20074732_chr15:20183050-20208457 hAstro.specific #1f78b4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>bedpe <span class="ot">&lt;-</span> astro_specific[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>, <span class="dv">12</span>,<span class="dv">11</span>)]</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co">#write.table(bedpe, "hAstro_sp_chr15_studentsdataframet.txt", sep = "\t", quote = F)</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>GB_specific <span class="ot">&lt;-</span> data[data<span class="sc">$</span>annotation <span class="sc">==</span> <span class="st">"GB.specific"</span>, ]</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(GB_specific)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       chr.x  start.x    end.x chr.y  start.y    end.y hAstro U3028 loopWidth
184977 chr15 17435126 17441610 chr15 18230127 18231848      0     2    792619
184978 chr15 17457775 17460493 chr15 18230127 18231848      0     2    771853
184985 chr15 18465420 18523463 chr15 19167038 19170276      0     1    674215
184988 chr15 18754402 18770002 chr15 19070204 19108832      0     1    327316
185091 chr15 22491925 22497810 chr15 22836947 22842739      0     3    344975
185092 chr15 22491925 22497810 chr15 23560790 23572771      0     3   1071913
          mango.FDR      mango.P
184977 5.180832e-08 1.517130e-08
184978 2.378743e-05 1.058543e-05
184985 3.138288e-05 1.445453e-05
184988 1.265685e-04 6.808408e-05
185091 0.000000e+00 0.000000e+00
185092 2.242247e-08 6.242519e-09
                                                loopID  annotation  colour
184977 chr15:17435126-17441610_chr15:18230127-18231848 GB.specific #e31a1c
184978 chr15:17457775-17460493_chr15:18230127-18231848 GB.specific #e31a1c
184985 chr15:18465420-18523463_chr15:19167038-19170276 GB.specific #e31a1c
184988 chr15:18754402-18770002_chr15:19070204-19108832 GB.specific #e31a1c
185091 chr15:22491925-22497810_chr15:22836947-22842739 GB.specific #e31a1c
185092 chr15:22491925-22497810_chr15:23560790-23572771 GB.specific #e31a1c</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>bedpe <span class="ot">&lt;-</span> GB_specific[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>, <span class="dv">12</span>,<span class="dv">11</span>)]</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co">#write.table(bedpe, "GB_sp_chr15_studentsdataframet.txt", sep = "\t", quote = F)</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>common <span class="ot">&lt;-</span> data[data<span class="sc">$</span>annotation <span class="sc">==</span> <span class="st">"common"</span>, ]</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(common)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       chr.x  start.x    end.x chr.y  start.y    end.y hAstro U3028 loopWidth
185098 chr15 22783892 22796574 chr15 22836947 22842739      1    13     49610
185108 chr15 23530907 23535835 chr15 25044024 25085981      2     1   1531631
185109 chr15 23560790 23572771 chr15 23642154 23650027      1     5     79310
185135 chr15 24854175 24860816 chr15 24953757 24958824      1     1     98795
185138 chr15 24953757 24958824 chr15 25044024 25085981      1     3    108712
185147 chr15 25044024 25085981 chr15 25432124 25440977      1     2    371548
          mango.FDR      mango.P
185098 0.000000e+00 0.000000e+00
185108 8.549782e-09 2.249694e-09
185109 3.846195e-06 1.385685e-06
185135 0.000000e+00 0.000000e+00
185138 0.000000e+00 0.000000e+00
185147 0.000000e+00 0.000000e+00
                                                loopID annotation  colour
185098 chr15:22783892-22796574_chr15:22836947-22842739     common #a0a0a0
185108 chr15:23530907-23535835_chr15:25044024-25085981     common #a0a0a0
185109 chr15:23560790-23572771_chr15:23642154-23650027     common #a0a0a0
185135 chr15:24854175-24860816_chr15:24953757-24958824     common #a0a0a0
185138 chr15:24953757-24958824_chr15:25044024-25085981     common #a0a0a0
185147 chr15:25044024-25085981_chr15:25432124-25440977     common #a0a0a0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>bedpe <span class="ot">&lt;-</span> common[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>, <span class="dv">12</span>,<span class="dv">11</span>)]</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co">#write.table(bedpe, "common_chr15_studentsdataframet.txt", sep = "\t", quote = F)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For demonstration sake I am going to convert the GB_sp_chr15_studentsdataframe.txt to a .bedpe file in shell, but you can use any of them for practise, for comparison, etc.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#less GB_sp_chr15_studentsdataframet.txt</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co">#cat GB_sp_chr15_studentsdataframet.txt | wc -l #2424</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#tail -n +2 GB_sp_chr15_studentsdataframet.txt &gt; tmpfile</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">#cut -f 2-9 tmpfile &gt; GB_sp_chr15_studentsdataframet.bedpe</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#less GB_sp_chr15_studentsdataframet.bedpe</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">#cat GB_sp_chr15_studentsdataframet.bedpe | wc -l #2423</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we will use Genomic Interactions and the UCSC’s genomic annotations for hg38 to annoate the genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Run Genomic annotations</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>dfloops <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"~/Desktop/PALS2024_3Dcourse/PALS2024/GB_sp_chr15_studentsdataframet.bedpe"</span>) <span class="co">#change path</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>dfinter <span class="ot">&lt;-</span> <span class="fu">makeGenomicInteractionsFromFile</span>(dfloops, <span class="at">type =</span> <span class="st">'bedpe'</span>, <span class="at">experiment_name =</span> <span class="st">"GB specific"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in makeGenomicInteractionsFromFile(dfloops, type = "bedpe",
experiment_name = "GB specific"): Some counts are set to zero, bedpe score
field may represent other data</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">interactionCounts</span>(dfinter))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.517130e-08 1.058543e-05 1.445453e-05 6.808408e-05 0.000000e+00
[6] 6.242519e-09</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>median_distance_interactions <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">calculateDistances</span>(dfinter, <span class="at">method =</span> <span class="st">"midpoint"</span>))</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>median_distance_interactions</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 335226</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> TxDb.Hsapiens.UCSC.hg38.knownGene</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>refseq.genes <span class="ot">&lt;-</span> <span class="fu">genes</span>(TxDb.Hsapiens.UCSC.hg38.knownGene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>  2135 genes were dropped because they have exons located on both strands
  of the same reference sequence or on more than one reference sequence,
  so cannot be represented by a single genomic range.
  Use 'single.strand.genes.only=FALSE' to get all the genes in a
  GRangesList object, or use suppressMessages() to suppress this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>refseq.transcripts <span class="ot">=</span> <span class="fu">transcriptsBy</span>(TxDb.Hsapiens.UCSC.hg38.knownGene, <span class="at">by=</span><span class="st">"gene"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>non_pseudogene <span class="ot">=</span> <span class="fu">names</span>(refseq.transcripts) <span class="sc">%in%</span> <span class="fu">unlist</span>(refseq.genes<span class="sc">$</span>gene_id) </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>refseq.transcripts <span class="ot">=</span> refseq.transcripts[non_pseudogene] </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>refseq.promoters <span class="ot">=</span> <span class="fu">promoters</span>(refseq.transcripts, <span class="at">upstream=</span><span class="dv">2000</span>, <span class="at">downstream=</span><span class="dv">2000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 2 out-of-bound ranges located on sequences
  chr4_GL000257v2_alt and chr16_GL383556v1_alt. Note that ranges located
  on a sequence whose length is unknown (NA) or on a circular sequence
  are not considered out-of-bound (use seqlengths() and isCircular() to
  get the lengths and circularity flags of the underlying sequences). You
  can use trim() to trim these ranges. See ?`trim,GenomicRanges-method`
  for more information.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># unlist object so "strand" is one vector</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>refseq.transcripts.ul <span class="ot">=</span> <span class="fu">unlist</span>(refseq.transcripts) </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># terminators can be called as promoters with the strand reversed</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">strand</span>(refseq.transcripts.ul) <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">strand</span>(refseq.transcripts.ul) <span class="sc">==</span> <span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>) </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>refseq.terminators.ul <span class="ot">=</span> <span class="fu">promoters</span>(refseq.transcripts.ul, <span class="at">upstream=</span><span class="dv">1000</span>, <span class="at">downstream=</span><span class="dv">1000</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 1 out-of-bound range located on sequence
  chr8_KI270815v1_alt. Note that ranges located on a sequence whose
  length is unknown (NA) or on a circular sequence are not considered
  out-of-bound (use seqlengths() and isCircular() to get the lengths and
  circularity flags of the underlying sequences). You can use trim() to
  trim these ranges. See ?`trim,GenomicRanges-method` for more
  information.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># change back to original strand</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">strand</span>(refseq.terminators.ul) <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">strand</span>(refseq.terminators.ul) <span class="sc">==</span> <span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>) </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="co"># `relist' maintains the original names and structure of the list</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>refseq.terminators <span class="ot">=</span> <span class="fu">relist</span>(refseq.terminators.ul, refseq.transcripts)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>annotation.features <span class="ot">=</span> <span class="fu">list</span>(<span class="at">promoter=</span>refseq.promoters, </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>                           <span class="at">terminator=</span>refseq.terminators, </span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>                           <span class="at">gene.body=</span>refseq.transcripts)</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a><span class="fu">annotateInteractions</span>(dfinter, annotation.features)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Annotating with promoter ...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Annotating with terminator ...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Annotating with gene.body ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">annotationFeatures</span>(dfinter)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "distal"     "promoter"   "gene.body"  "terminator"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">categoriseInteractions</span>(dfinter)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                category count
1          distal-distal    26
2        distal-promoter   316
3       distal-gene.body    21
4      distal-terminator    12
5      promoter-promoter  1530
6     promoter-gene.body   346
7    promoter-terminator   132
8    gene.body-gene.body    29
9   gene.body-terminator     9
10 terminator-terminator     2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>dfinter_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(dfinter)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dfinter_df[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">6</span>,<span class="dv">10</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">15</span>,<span class="dv">19</span>,<span class="dv">20</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  seqnames1   start1     end1 node.class1 seqnames2   start2     end2
1     chr15 17435127 17441610      distal     chr15 18230128 18231848
2     chr15 17457776 17460493      distal     chr15 18230128 18231848
3     chr15 18465421 18523463      distal     chr15 19167039 19170276
4     chr15 18754403 18770002      distal     chr15 19070205 19108832
5     chr15 22491926 22497810      distal     chr15 22836948 22842739
6     chr15 22491926 22497810      distal     chr15 23560791 23572771
  node.class2       counts                                            name
1      distal 1.517130e-08 chr15:17435126-17441610_chr15:18230127-18231848
2      distal 1.058543e-05 chr15:17457775-17460493_chr15:18230127-18231848
3      distal 1.445453e-05 chr15:18465420-18523463_chr15:19167038-19170276
4      distal 6.808408e-05 chr15:18754402-18770002_chr15:19070204-19108832
5    promoter 0.000000e+00 chr15:22491925-22497810_chr15:22836947-22842739
6    promoter 6.242519e-09 chr15:22491925-22497810_chr15:23560790-23572771</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>dfinter_df<span class="sc">$</span>start1 <span class="ot">&lt;-</span> dfinter_df<span class="sc">$</span>start1 <span class="sc">-</span><span class="dv">1</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>dfinter_df<span class="sc">$</span>start2 <span class="ot">&lt;-</span> dfinter_df<span class="sc">$</span>start2 <span class="sc">-</span><span class="dv">1</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dfinter_df[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">6</span>,<span class="dv">10</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">15</span>,<span class="dv">19</span>,<span class="dv">20</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  seqnames1   start1     end1 node.class1 seqnames2   start2     end2
1     chr15 17435126 17441610      distal     chr15 18230127 18231848
2     chr15 17457775 17460493      distal     chr15 18230127 18231848
3     chr15 18465420 18523463      distal     chr15 19167038 19170276
4     chr15 18754402 18770002      distal     chr15 19070204 19108832
5     chr15 22491925 22497810      distal     chr15 22836947 22842739
6     chr15 22491925 22497810      distal     chr15 23560790 23572771
  node.class2       counts                                            name
1      distal 1.517130e-08 chr15:17435126-17441610_chr15:18230127-18231848
2      distal 1.058543e-05 chr15:17457775-17460493_chr15:18230127-18231848
3      distal 1.445453e-05 chr15:18465420-18523463_chr15:19167038-19170276
4      distal 6.808408e-05 chr15:18754402-18770002_chr15:19070204-19108832
5    promoter 0.000000e+00 chr15:22491925-22497810_chr15:22836947-22842739
6    promoter 6.242519e-09 chr15:22491925-22497810_chr15:23560790-23572771</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#here we are annotating the loops as promoter-promoter (pp), when both anchors are annotated to promoters, promoter-enhancer, when a promoter anchor is connected to an anchor which is not a promoter, we call distal-distal - (intergenic/intronic regions connected to anchors) as enhancer-enhancer (ee loops). Usually gene-body (exonic) or terminator (20bp GC rich region) are marked with specific histone marks such as H3K36Me3 at exons, what their interactions with each other or to a distal region inplies is not very clear so we call such interactions 'others'.</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>dfinter_df <span class="ot">&lt;-</span> dfinter_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">interaction.type =</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>                                             <span class="fu">case_when</span>((dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"promoter"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"promoter"</span><span class="sc">~</span><span class="st">"pp"</span>),</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"promoter"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"distal"</span><span class="sc">~</span><span class="st">"pe"</span>),</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"distal"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"promoter"</span><span class="sc">~</span><span class="st">"pe"</span>),</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"promoter"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"gene.body"</span><span class="sc">~</span><span class="st">"pe"</span>),</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"gene.body"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"promoter"</span><span class="sc">~</span><span class="st">"pe"</span>),</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"promoter"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"terminator"</span><span class="sc">~</span><span class="st">"pe"</span>),</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"terminator"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"promoter"</span><span class="sc">~</span><span class="st">"pe"</span>),</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"distal"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"distal"</span><span class="sc">~</span><span class="st">"ee"</span>),</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"distal"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"gene.body"</span><span class="sc">~</span><span class="st">"others"</span>),</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"gene.body"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"distal"</span><span class="sc">~</span><span class="st">"others"</span>),</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"gene.body"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"gene.body"</span><span class="sc">~</span><span class="st">"others"</span>),</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"distal"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"terminator"</span><span class="sc">~</span><span class="st">"others"</span>),</span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"terminator"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"distal"</span><span class="sc">~</span><span class="st">"others"</span>),</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"terminator"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"terminator"</span><span class="sc">~</span><span class="st">"others"</span>),</span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"gene.body"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"terminator"</span><span class="sc">~</span><span class="st">"others"</span>),</span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>                                                       (dfinter_df<span class="sc">$</span>node.class1 <span class="sc">==</span> <span class="st">"terminator"</span> <span class="sc">&amp;</span>dfinter_df<span class="sc">$</span>node.class2 <span class="sc">==</span> <span class="st">"gene.body"</span><span class="sc">~</span><span class="st">"others"</span>)))</span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a>dfinter_df_sh <span class="ot">&lt;-</span> dfinter_df[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">6</span>,<span class="dv">10</span>,<span class="dv">11</span>,<span class="dv">12</span>,<span class="dv">15</span>,<span class="dv">19</span>,<span class="dv">20</span>,<span class="dv">21</span>)]</span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dfinter_df_sh)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  seqnames1   start1     end1 node.class1 seqnames2   start2     end2
1     chr15 17435126 17441610      distal     chr15 18230127 18231848
2     chr15 17457775 17460493      distal     chr15 18230127 18231848
3     chr15 18465420 18523463      distal     chr15 19167038 19170276
4     chr15 18754402 18770002      distal     chr15 19070204 19108832
5     chr15 22491925 22497810      distal     chr15 22836947 22842739
6     chr15 22491925 22497810      distal     chr15 23560790 23572771
  node.class2       counts                                            name
1      distal 1.517130e-08 chr15:17435126-17441610_chr15:18230127-18231848
2      distal 1.058543e-05 chr15:17457775-17460493_chr15:18230127-18231848
3      distal 1.445453e-05 chr15:18465420-18523463_chr15:19167038-19170276
4      distal 6.808408e-05 chr15:18754402-18770002_chr15:19070204-19108832
5    promoter 0.000000e+00 chr15:22491925-22497810_chr15:22836947-22842739
6    promoter 6.242519e-09 chr15:22491925-22497810_chr15:23560790-23572771
  interaction.type
1               ee
2               ee
3               ee
4               ee
5               pe
6               pe</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">#write.csv(dfinter_df_sh, "GB_sp_chr15_GA.csv")</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co">#get genes annotated to the anchors of the loops</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>dfinter_pp <span class="ot">&lt;-</span> dfinter[<span class="fu">is.pp</span>(dfinter)]</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>dfinter_pp_genes <span class="ot">&lt;-</span> dfinter_pp<span class="sc">@</span>regions<span class="sc">@</span>elementMetadata<span class="sc">@</span>listData<span class="sc">$</span>promoter.id</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>dfinter_pp_genes_un <span class="ot">&lt;-</span> <span class="fu">unlist</span>(dfinter_pp_genes)</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>dfinter_pp_genes_un <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(dfinter_pp_genes_un)</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>dfinter_pp_genes_dedup <span class="ot">&lt;-</span> dfinter_pp_genes_un[<span class="sc">!</span><span class="fu">duplicated</span>(dfinter_pp_genes_un)]</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>converted <span class="ot">&lt;-</span> <span class="fu">bitr</span>(dfinter_pp_genes_dedup, <span class="at">fromType =</span> <span class="st">"ENTREZID"</span>, <span class="at">toType =</span> <span class="fu">c</span>(<span class="st">"ENSEMBL"</span>, <span class="st">"SYMBOL"</span>), <span class="at">OrgDb =</span> org.Hs.eg.db, <span class="at">drop =</span> <span class="cn">FALSE</span>) <span class="co">#14054</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(dfinter_pp_genes_dedup, fromType = "ENTREZID", toType =
c("ENSEMBL", : 11.25% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>converted_pp <span class="ot">&lt;-</span> converted</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>converted_pp<span class="sc">$</span>annotation <span class="ot">&lt;-</span> <span class="st">"pp"</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>converted_pp <span class="ot">&lt;-</span> converted_pp[<span class="sc">!</span><span class="fu">duplicated</span>(converted_pp<span class="sc">$</span>ENTREZID),] </span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>dfinter_pd <span class="ot">&lt;-</span> dfinter[<span class="fu">is.pd</span>(dfinter)]</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>dfinter_pd_genes <span class="ot">&lt;-</span> dfinter_pd<span class="sc">@</span>regions<span class="sc">@</span>elementMetadata<span class="sc">@</span>listData<span class="sc">$</span>promoter.id</span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>dfinter_pd_genes_un <span class="ot">&lt;-</span> <span class="fu">unlist</span>(dfinter_pd_genes)</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>dfinter_pd_genes_un <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(dfinter_pd_genes_un)</span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>dfinter_pd_genes_dedup <span class="ot">&lt;-</span> dfinter_pd_genes_un[<span class="sc">!</span><span class="fu">duplicated</span>(dfinter_pd_genes_un)] </span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>converted_pd <span class="ot">&lt;-</span> <span class="fu">bitr</span>(dfinter_pd_genes_dedup, <span class="at">fromType =</span> <span class="st">"ENTREZID"</span>, <span class="at">toType =</span> <span class="fu">c</span>(<span class="st">"ENSEMBL"</span>, <span class="st">"SYMBOL"</span>), <span class="at">OrgDb =</span> org.Hs.eg.db, <span class="at">drop =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(dfinter_pd_genes_dedup, fromType = "ENTREZID", toType =
c("ENSEMBL", : 11.25% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>dfinter_pt <span class="ot">&lt;-</span> dfinter[<span class="fu">is.pt</span>(dfinter)]</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>dfinter_pt_genes <span class="ot">&lt;-</span> dfinter_pt<span class="sc">@</span>regions<span class="sc">@</span>elementMetadata<span class="sc">@</span>listData<span class="sc">$</span>promoter.id</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>dfinter_pt_genes_un <span class="ot">&lt;-</span> <span class="fu">unlist</span>(dfinter_pt_genes)</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>dfinter_pt_genes_un <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(dfinter_pt_genes_un)</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>dfinter_pt_genes_dedup <span class="ot">&lt;-</span> dfinter_pt_genes_un[<span class="sc">!</span><span class="fu">duplicated</span>(dfinter_pt_genes_un)] </span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>converted_pt <span class="ot">&lt;-</span> <span class="fu">bitr</span>(dfinter_pt_genes_dedup, <span class="at">fromType =</span> <span class="st">"ENTREZID"</span>, <span class="at">toType =</span> <span class="fu">c</span>(<span class="st">"ENSEMBL"</span>, <span class="st">"SYMBOL"</span>), <span class="at">OrgDb =</span> org.Hs.eg.db, <span class="at">drop =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(dfinter_pt_genes_dedup, fromType = "ENTREZID", toType =
c("ENSEMBL", : 11.25% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>dfinter_pg <span class="ot">&lt;-</span> dfinter[<span class="fu">isInteractionType</span>(dfinter, <span class="st">"promoter"</span>, <span class="st">"gene.body"</span>)]</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>dfinter_pg_genes <span class="ot">&lt;-</span> dfinter_pg<span class="sc">@</span>regions<span class="sc">@</span>elementMetadata<span class="sc">@</span>listData<span class="sc">$</span>promoter.id</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>dfinter_pg_genes_un <span class="ot">&lt;-</span> <span class="fu">unlist</span>(dfinter_pg_genes)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>dfinter_pg_genes_un <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(dfinter_pg_genes_un)</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>dfinter_pg_genes_dedup <span class="ot">&lt;-</span> dfinter_pg_genes_un[<span class="sc">!</span><span class="fu">duplicated</span>(dfinter_pg_genes_un)] <span class="co">#2715</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>converted_pg <span class="ot">&lt;-</span> <span class="fu">bitr</span>(dfinter_pg_genes_dedup, <span class="at">fromType =</span> <span class="st">"ENTREZID"</span>, <span class="at">toType =</span> <span class="fu">c</span>(<span class="st">"ENSEMBL"</span>, <span class="st">"SYMBOL"</span>), <span class="at">OrgDb =</span> org.Hs.eg.db, <span class="at">drop =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(dfinter_pg_genes_dedup, fromType = "ENTREZID", toType =
c("ENSEMBL", : 11.25% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>converted_pe <span class="ot">&lt;-</span> <span class="fu">rbind</span>(converted_pd, converted_pg, converted_pt)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>converted_pe <span class="ot">&lt;-</span> converted_pe[<span class="sc">!</span><span class="fu">duplicated</span>(converted_pe<span class="sc">$</span>ENTREZID),]</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>converted_pe<span class="sc">$</span>annotation <span class="ot">&lt;-</span> <span class="st">"pe"</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>converted <span class="ot">&lt;-</span> <span class="fu">rbind</span>(converted_pp, converted_pe) </span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="co">#unique genes annotated to anchors </span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(converted<span class="sc">$</span>ENTREZID)) <span class="co">#[1] 622 In this case all gene promoters that are annotated to pp loops are also annotated to pe</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 622</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co">#write.csv(converted, "GB_sp_chr15_GA_genes.csv")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="analysing-lengths-of-differential-loops" class="level4">
<h4 class="anchored" data-anchor-id="analysing-lengths-of-differential-loops">Analysing lengths of differential loops</h4>
<p>Loop lengths help us understand how the chromatin is folded in a specific ondition with respect to others. Here we are looking how U3028 GB specifc loops are compacted in chr 15 in comparison to loops that are unchanged in hAstro and GB (common), and control hAstro specific loops.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co">#we will look at the median and interquartile ranges in our analysis. You can also look at mean and standard error</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>data_summary <span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="fu">quantile</span>(x, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>))</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(out) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ymin"</span>,<span class="st">"y"</span>,<span class="st">"ymax"</span>)</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(out) </span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>log2.loopWidth <span class="ot">&lt;-</span> <span class="fu">log2</span>(data<span class="sc">$</span>loopWidth)</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a><span class="co">#look at the distribution of loop length</span></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(data<span class="sc">$</span>loopWidth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(data<span class="sc">$</span>log2.loopWidth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co">#since in our case the data does not look normal, we shall use Wilcoxon test to look at statistical significance in loop length</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>annotation <span class="ot">&lt;-</span> <span class="fu">factor</span>(data<span class="sc">$</span>annotation, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"hAstro.specific"</span>, <span class="st">"GB.specific"</span>, <span class="st">"common"</span>) )</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>stat.test1 <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wilcox_test</span>(log2.loopWidth<span class="sc">~</span>annotation, <span class="at">p.adjust.method =</span> <span class="st">"BH"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_significance</span>()</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>stat.test1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 9
  .y.         group1 group2    n1    n2 statistic        p    p.adj p.adj.signif
  &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;       
1 log2.loopW… hAstr… GB.sp…   426  2423   490788  1.06e- 1 1.06e- 1 ns          
2 log2.loopW… hAstr… common   426   662   202964. 1.74e-34 2.61e-34 ****        
3 log2.loopW… GB.sp… common  2423   662  1194848. 2.39e-83 7.17e-83 ****        </code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>stat.test1 <span class="ot">&lt;-</span> stat.test1 <span class="sc">%&gt;%</span> <span class="fu">add_xy_position</span>(<span class="at">x =</span> <span class="st">"annotation"</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>stattest1df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(stat.test1)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>stattest1df<span class="sc">$</span>groups <span class="ot">&lt;-</span> <span class="fu">as.character</span>(stattest1df<span class="sc">$</span>groups)</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="co">#save the statical data frame for article and replotting if needed</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a><span class="co">#write.csv(stattest1df, "Wilcoxon.test.csv")</span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a><span class="co">#here build the plot step wise</span></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>data<span class="sc">$</span>annotation, <span class="at">y=</span>data<span class="sc">$</span>loopWidth)) <span class="sc">+</span> <span class="fu">geom_violin</span>(<span class="at">fill =</span> <span class="st">"gray"</span>, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">linewidth=</span><span class="fl">0.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>p1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-8-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p1 <span class="sc">+</span> <span class="fu">stat_summary</span>(<span class="at">fun.data=</span>data_summary, <span class="at">size =</span> <span class="fl">0.1</span>)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-8-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p2 <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(), <span class="at">panel.grid.minor.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>p3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-8-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> p3  <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Loop length"</span>) <span class="sc">+</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">size=</span> <span class="dv">10</span>)) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Annotations"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"loop length (bp)"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">6</span>, <span class="at">angle =</span> <span class="dv">0</span>),</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>                                                                                                                          <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>),<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>                                                                                                                          <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>),  <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>))</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>p4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-8-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> p4 <span class="sc">+</span> <span class="fu">stat_pvalue_manual</span>(</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>  stat.test1, <span class="at">label =</span> <span class="st">"Wilcoxon-test, p.adj = {p.adj} {p.adj.signif}"</span>,</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">vjust =</span> <span class="sc">-</span><span class="fl">1.3</span>, <span class="at">bracket.nudge.y =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">1.5</span>,<span class="at">y.position =</span> <span class="fu">seq</span>(<span class="at">from=</span><span class="dv">2050000</span>,<span class="at">by=</span><span class="dv">400000</span>,<span class="at">length.out=</span><span class="dv">3</span>)</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="dv">5000</span>, <span class="fu">max</span>(<span class="fu">seq</span>(<span class="at">from=</span><span class="dv">2050000</span>,<span class="at">by=</span><span class="dv">400000</span>,<span class="at">length.out=</span><span class="dv">4</span>)))</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>p5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="loop-additional-analysis_files/figure-html/unnamed-chunk-8-7.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co">#png("looplength.png", width = 4, height = 6, res = 600, units = 'in')</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co">#p5</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="co">#par(mar=c(8, 6, 6, 5) + 0.1)</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="co">#dev.off()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="extracting-anchors-from-loops-for-more-downstream-analysis" class="level4">
<h4 class="anchored" data-anchor-id="extracting-anchors-from-loops-for-more-downstream-analysis">Extracting anchors from loops for more downstream analysis</h4>
<p>In this part of the analysis, we might want to overlap anchors to ChIP-Seq peaks obtained for TFs, histone marks, CTCF etc., to integrate loops to yopur epigenetic data to understand which epigenetic regions are re organized 3D.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co">#fetch anchors for downstream analysis</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="co">#for GB specific loops</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co">#anchor.x</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>GB_specific <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(annotation <span class="sc">==</span>  <span class="st">"GB.specific"</span>)</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>anchor.x <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> GB_specific<span class="sc">$</span>chr.x, <span class="fu">IRanges</span>(<span class="at">start =</span> GB_specific<span class="sc">$</span>start.x, <span class="at">end =</span>GB_specific<span class="sc">$</span>end.x))</span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a>anchor.x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2423 ranges and 0 metadata columns:
         seqnames              ranges strand
            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
     [1]    chr15   17435126-17441610      *
     [2]    chr15   17457775-17460493      *
     [3]    chr15   18465420-18523463      *
     [4]    chr15   18754402-18770002      *
     [5]    chr15   22491925-22497810      *
     ...      ...                 ...    ...
  [2419]    chr15 101292364-101299094      *
  [2420]    chr15 101292364-101299094      *
  [2421]    chr15 101307065-101336013      *
  [2422]    chr15 101389337-101395268      *
  [2423]    chr15 101648802-101656417      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>anchor.y <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> GB_specific<span class="sc">$</span>chr.y, <span class="fu">IRanges</span>(<span class="at">start =</span> GB_specific<span class="sc">$</span>start.y, <span class="at">end =</span>GB_specific<span class="sc">$</span>end.y))</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>anchor.y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2423 ranges and 0 metadata columns:
         seqnames              ranges strand
            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
     [1]    chr15   18230127-18231848      *
     [2]    chr15   18230127-18231848      *
     [3]    chr15   19167038-19170276      *
     [4]    chr15   19070204-19108832      *
     [5]    chr15   22836947-22842739      *
     ...      ...                 ...    ...
  [2419]    chr15 101722364-101725660      *
  [2420]    chr15 101953134-101970476      *
  [2421]    chr15 101648802-101656417      *
  [2422]    chr15 101608901-101621083      *
  [2423]    chr15 101722364-101725660      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>bed_file <span class="ot">&lt;-</span> <span class="fu">GRangesList</span>(anchor.x, anchor.y)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>bed_file <span class="ot">&lt;-</span> <span class="fu">unlist</span>(bed_file)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>bed_file</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 4846 ranges and 0 metadata columns:
         seqnames              ranges strand
            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
     [1]    chr15   17435126-17441610      *
     [2]    chr15   17457775-17460493      *
     [3]    chr15   18465420-18523463      *
     [4]    chr15   18754402-18770002      *
     [5]    chr15   22491925-22497810      *
     ...      ...                 ...    ...
  [4842]    chr15 101722364-101725660      *
  [4843]    chr15 101953134-101970476      *
  [4844]    chr15 101648802-101656417      *
  [4845]    chr15 101608901-101621083      *
  [4846]    chr15 101722364-101725660      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>bed_unique <span class="ot">&lt;-</span> <span class="fu">unique</span>(bed_file)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>bed_unique</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 789 ranges and 0 metadata columns:
        seqnames              ranges strand
           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
    [1]    chr15   17435126-17441610      *
    [2]    chr15   17457775-17460493      *
    [3]    chr15   18465420-18523463      *
    [4]    chr15   18754402-18770002      *
    [5]    chr15   22491925-22497810      *
    ...      ...                 ...    ...
  [785]    chr15 101608901-101621083      *
  [786]    chr15 101722364-101725660      *
  [787]    chr15 101751634-101768881      *
  [788]    chr15 101595158-101600078      *
  [789]    chr15 101953134-101970476      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co">#export.bed(bed_unique,"GB.specific.anchors.bed")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>