# Downstream-analysis-3D-chromatin-loops
Additional analysis to understand chromatin loop impact in cells

A short tutorial on R to integrate loops from HiC/ HiChIP for further downstream analysis.

## Example analysis
Use the requiredPackages.R to install the packages we shall use in the analysis. The entire tutorial with output is depicted in an HTML file called loop-additional-analysis.html in the PALS2024 folder.

Download the tutorial with git

```
git clone https://github.com/remeseirogrp/Downstream-analysis-3D-chromatin-loops.git

```

## The additional_files folder
This folder has the raw script and the raw data that was used to build the tutorial. Example output for UPSET plot, loop length and other output files are located in this folder.

## Web resources for publicly available 3D data and tools

[4D Nucleome](https://www.4dnucleome.org/)

[3D Genome Browser](http://3dgenome.fsm.northwestern.edu/)

[Hemtools for HiC and capture C](https://hemtools.readthedocs.io/en/latest/content/NGS_pipelines/hicpro_split.html)

[3D-genome](https://github.com/topics/3d-genome)

[HiC-analysis with HiCExplorer with GalaxyTools](https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/hicexplorer/tutorial.html)

[HiC-Pro](https://github.com/nservant/HiC-Pro)

[HiC tools: Aiden Lab](https://github.com/aidenlab)

[HiC/ HiChIP tools: Ferhat Ay lab](https://github.com/ay-lab)

[HiC/ HiChIP tools: Aryee lab](https://github.com/aryeelab)

## Articles

**Methods for mapping 3D chromosome architecture**

[Kempfer, R., Pombo, A. Methods for mapping 3D chromosome architecture. Nat Rev Genet 21, 207–226 (2020). https://doi.org/10.1038/s41576-019-0195-2](https://www.nature.com/articles/s41576-019-0195-2)

**Understanding 3D genome organization by multidisciplinary methods**

[https://www.nature.com/articles/s41580-021-00362-w](https://www.nature.com/articles/s41580-021-00362-w)

**The era of 3D and spatial genomics**

[Bouwman BAM, Crosetto N, Bienko M. The era of 3D and spatial genomics. Trends Genet. 2022 Oct;38(10):1062-1075. doi: 10.1016/j.tig.2022.05.010. Epub 2022 Jun 6. PMID: 35680466.](https://www.cell.com/trends/genetics/fulltext/S0168-9525(22)00118-4)

**Epigenetics, Enhancer Function and 3D Chromatin Organization in Reprogramming to Pluripotency**

[Hörnblad A, Remeseiro S. Epigenetics, Enhancer Function and 3D Chromatin Organization in Reprogramming to Pluripotency. Cells. 2022; 11(9):1404. https://doi.org/10.3390/cells11091404 ](https://www.mdpi.com/2073-4409/11/9/1404)

**Computational methods for analyzing multiscale 3D genome organization**

[Zhang, Y., Boninsegna, L., Yang, M. et al. Computational methods for analysing multiscale 3D genome organization. Nat Rev Genet 25, 123–141 (2024). https://doi.org/10.1038/s41576-023-00638-1](https://www.nature.com/articles/s41576-023-00638-1)

**Comparison of computational methods for 3D genome analysis at single-cell Hi-C level**

[Li X, An Z, Zhang Z. Comparison of computational methods for 3D genome analysis at single-cell Hi-C level. Methods. 2020 Oct 1;181-182:52-61. doi: 10.1016/j.ymeth.2019.08.005. Epub 2019 Aug 21. PMID: 31445093.](https://www.sciencedirect.com/science/article/pii/S1046202319300891)

**Computational methods for predicting 3D genomic organization from high-resolution chromosome conformation capture data**

[Kimberly MacKay, Anthony Kusalik, Computational methods for predicting 3D genomic organization from high-resolution chromosome conformation capture data, Briefings in Functional Genomics, Volume 19, Issue 4, July 2020, Pages 292–308, https://doi.org/10.1093/bfgp/elaa004](https://academic.oup.com/bfg/article/19/4/292/5826457)





